test/Push.hs

```diff
@@ -0,0 +1,367 @@
+      -- differences such as ordering and spacing in the JSON Aeson generates
+      property $ \(NotificationDecoding (bs, notification)) ->
+        (A.encode notification) ==
+        (A.encode . fromJust . (A.decode :: ByteString -> Maybe Notification) $
+         bs)
+
+-- A single example of a FCM notification which we expect to aeson decode to
+-- the paired Notification value.
+validFCMDecoding :: (B.ByteString, Notification)
+validFCMDecoding =
+  let bs =
+        B.unlines
+          [ "{"
+          , "\"interests\": [\"interestOne\"],"
+          , "\"fcm\": { \"notification\": {"
+          , "              \"body\": \"bodyText\","
+          , "              \"title\": \"titleText\""
+          , "                             }"
+          , "         }"
+          , "}"
+          ]
+      notification =
+        Notification
+        { notificationInterest = fromJust . mkInterest $ "interestOne"
+        , notificationWebhookURL = Nothing
+        , notificationWebhookLevel = Nothing
+        , notificationAPNSPayload = Nothing
+        , notificationGCMPayload = Nothing
+        , notificationFCMPayload =
+            Just $
+            FCMPayload $
+            HM.fromList
+              [ ("notification" :: T.Text) .=
+                (A.Object $
+                 HM.fromList
+                   [ ("title" :: T.Text) .= ("titleText" :: T.Text)
+                   , ("body" :: T.Text) .= ("bodyText" :: T.Text)
+                   ])
+              ]
+        }
+  in (bs, notification)
+
+-- | A ByteString will decode to some value 'a'.
+type Decoding a = (B.ByteString, a)
+
+newtype NotificationDecoding =
+  NotificationDecoding (Decoding Notification)
+  deriving (Show)
+
+newtype APNSDecoding =
+  APNSDecoding (Decoding (Maybe APNSPayload))
+
+newtype FCMDecoding =
+  FCMDecoding (Decoding (Maybe FCMPayload))
+
+newtype GCMDecoding =
+  GCMDecoding (Decoding (Maybe GCMPayload))
+
+instance Arbitrary APNSDecoding where
+  arbitrary = do
+    InterestText titleText <- arbitrary
+    InterestText bodyText <- arbitrary
+    NonRecJSON dataJSON <- arbitrary
+    let alert :: A.Object
+        alert = HM.fromList ["title" .= titleText, "body" .= bodyText]
+        aps :: A.Object
+        aps = HM.fromList ["alert" .= alert]
+        payload :: A.Object
+        payload = HM.fromList ["aps" .= aps, "data" .= dataJSON]
+    let apns = APNSPayload payload
+        bs :: ByteString
+        bs =
+          mconcat
+            [ "{"
+            , "\"aps\":"
+            , "{"
+            , "              \"alert\":{\"title\":\""
+            , (B.pack . T.unpack $ titleText)
+            , "\""
+            , "                        ,\"body\":\""
+            , (B.pack . T.unpack $ bodyText)
+            , "\""
+            , "                        }"
+            , "            }"
+            , ",\"data\":"
+            , (A.encode dataJSON)
+            , "}"
+            ]
+    return $ APNSDecoding (bs, Just apns)
```

Comment:
`decode . encode === id` would only test that the decoder and encoder are inverses of each other. This is an expected property but since we don't need to decode Payloads as part of our API is not what we really want to test.

What I'm testing here is that the encoder produces expected output. In doing so I also test the `decode . encode === id` property because I use a decoding to 'normalise' the ByteString to Aesons field order.

I'm sure there might be better ways to do this.